onnxruntime/core/graph/contrib_ops/contrib_defs.cc

```diff
@@ -441,7 +441,98 @@ and op)DOC";
       .Attr("scale", "A 1-D values of (height, width).", AttributeProto::INTS, OPTIONAL)
       .Input(0, "input", "Input tensor of shape [N,C,H,W]", "T")
       .Output(0, "output", "Result, has same type as input, with H and W dimensions reduced.", "T")
-      .TypeConstraint("T", {"tensor(float16)", "tensor(float)", "tensor(double)"}, "Constrain input and output types to float tensors.");
+      .TypeConstraint("T", {"tensor(float16)", "tensor(float)", "tensor(double)"}, "Constrain input and output types to float tensors.")
+      .TypeAndShapeInferenceFunction([](ONNX_NAMESPACE::InferenceContext& ctx) {
+        // Type inference
+        ONNX_NAMESPACE::propagateElemTypeFromInputToOutput(ctx, 0, 0);
+
+        // Shape inference
+        auto* output_shape =
+            ctx.getOutputType(0)->mutable_tensor_type()->mutable_shape();
+
+        if (ONNX_NAMESPACE::hasNInputShapes(ctx, 1)) {
+          const auto& input_shape =
+              ctx.getInputType(0)->tensor_type().shape();
+          const auto input_rank =
+              input_shape.dim_size();
+          if (input_rank != 4)
+            fail_shape_inference("Input's shape must be 4-D");
+
+          // parse necessary attributes for futher processing
+          std::vector<int64_t> border;
+          bool border_present =
+              getRepeatedAttribute(ctx, "border", border);
+          if (!border_present || border.size() != 4)
```

Comment:
Border seems to be an optional attribute as per the spec. Why do we fail if it's not present?

Reply:
There seems to be a disconnect somewhere - I don't see how processing can be performed without border accroding to the core operator logic (it may process with or without 'scale' attribute). In fact, even the op implementation has this check (which doesn't make sense if the op treats this attribute as optional) - 

https://github.com/microsoft/onnxruntime/blob/fa2eea73399c9625fa79761a4b84f240e0ec9669/onnxruntime/contrib_ops/cpu/crop.h#L22

I think the schema has a bug, only 'scale' was meant to be optional and 'border' was required but since the schema has been present for some time, I vote we keep it as such. Obviously, we cannot proceed in the implementation without 'border' though...

Reply:
If the implementation cannot proceed without border being present, then just mark it as mandatory in the spec. It cannot break existing models any way since the op implementation will check for it. Better to correct the spec and avoid confusion.

Reply:
Removed the "OPTIONAL" specifier for the attribute in the spec (this is a deprecated experimental op, so refining the spec here)